Rust examples

marvin-hansen opened the pull request:
As discussed with Alex Eagle on Slack recently,

this PR contributes multiple Rust examples that cover roughly the top ten use cases. 
All examples use the new Bazelmod format. The format follows loosely the great C++ tutorial.

If there is anything that needs adjustment, please let me know.

Formalities:
* I've signed the  Google CLA
* I've signed all commits

alexeagle commented on .gitignore:
```diff
@@ -10,3 +10,6 @@ target/
 # Xcode
 *.xcodeproj
 node_modules
+/rust-tutorial/*/MODULE.bazel.lock
```
I'd suggest you put these in a nested `.gitignore` file so future contributors under the rust folder don't need a root approver.

alexeagle commented on README.md:
```diff
@@ -37,6 +37,10 @@ Note that tutorials for other languages may be found under other repositories:
 
+### Rust
+
+* [Using Rust with Bzlmod](/rust-tutorial)
```
nit: "with Bzlmod" seems redundant, this whole repo is converted to Bazel 7 and bzlmod quite a while ago.

alexeagle commented on rust-tutorial/00-hello-world/.bazeliskrc:
```diff
@@ -0,0 +1 @@
+USE_BAZEL_VERSION=7.2.0
```
this should be in a `.bazelversion` file instead - tools other than bazelisk know to read that one.
Also, use symlinks in the PR so that there's only one copy of that file and the others point to it.

marvin-hansen replied:
Thanks. Didn't know that. Fixed that.

alexeagle commented on rust-tutorial/00-hello-world/MODULE.bazel:
```diff
@@ -0,0 +1,25 @@
+    version = "0.0.0"
+)
+###############################################################################
+# B A Z E L  C E N T R A L  R E G I S T R Y # https://registry.bazel.build/
```
this boundary looks very different from anything I've seen. I don't think any of the comments in this file are useful.

marvin-hansen replied:
Yeap. That was a a leftover from a template file. It's removed.

alexeagle commented on rust-tutorial/00-hello-world/BUILD.bazel:
Is there a reason not to use the root package for the code in this example?

marvin-hansen replied:
Good catch. Not at all. I've removed the redundant. stuff.

alexeagle reviewed (commented):
Will there be some prose to go along with this tutorial, like there is for C++ and Go and some others?

marvin-hansen commented:
I've found the source of the C++ and Go tutorial text in the Bazel main repo.

Just to clarify your question, do you ask for a tutorial similar to the C++ and fill a PR 
with the main repo to get a Rust getting started tutorial on the Bazel website?

In principal, I am okay with that, I just want to make sure I get this right?

alexeagle commented:
The Go tutorial isn't on the Bazel site: https://bazel-contrib.github.io/SIG-rules-authors/go-tutorial.html
In the past we have not figured out where this content belongs. Google doesn't staff the project with a tech writer currently so it may be better for tutorials to live in bazel-contrib or in the ruleset repo.

I'm asking about your future plans because you named these folders "tutorial", but without accompanying text they are not part of a tutorial. If they are just examples with a README that's fine too. Writing a real tutorial is a lot of work and also easily goes out-of-date so it's a future commitment as well.

marvin-hansen commented:
@alexeagle 

I see. Perfectly valid point. And you are right, this is not a tutorial similar to the Go one you shared; not even close. 
Sorry for the confusion, first time contribution to this project...

However, there  is no way I can allocate the time to put something like the Go tutorial together let alone maintain it over time.  This is a job for a professional tech writer, not a random contributor. Also, I am really unsure if that kind of overkill tutorial is actually needed because, in my observation, the Rust Bazel community tends to be filled with rather senior engineers and they may not even want a comprehensive first principle tutorial since they already know Bazel and only need to file some Rust specific gaps such as generating proto bindings, cross compiling, patching MUSL together etc.  And that's basically the gap I'm filling. 

I read so often comments in the rules_rust issue tracker that refer to, well, missing Rust Bazelmod examples. I actually [filled another PR over at the rules_rust repo](https://github.com/bazelbuild/rules_rust/pull/2713) to address at least half a dozen open Bazelmod issues that can be easily closed with a code example or a brief Readme. 
  
What I suggest to resolve this matter is to rename the folder to "rust-examples" and leave it there.

Wyverald commented on rust-tutorial/02-hello-cross/README.md:
```diff
@@ -0,0 +1,331 @@
+3) LLVM for ARM (aarch64)
+
+For the host LLVM, you just specify a LLVM version and then register the toolchain as usual. The target LLVM toolchains,
+however, have dependencies on system libraries for the target platform. Therefore, it is requires to download a so
```
```suggestion
however, have dependencies on system libraries for the target platform. Therefore, it is required to download a so-
```

Wyverald reviewed (changes requested):
thank you for your contribution! I've only looked at the first couple of examples so far, but will continue reviewing in a bit.

marvin-hansen commented on rust-tutorial/01-hello-clang/MODULE.bazel:
```diff
@@ -0,0 +1,55 @@
+)
+
+###############################################################################
+# B A Z E L  C E N T R A L  R E G I S T R Y # https://registry.bazel.build/
```
The comment header was contested before in the previous and in response, I've removed them all as it seems impossible to reach a consensus here. That's okay. Just FYI, the blueprint for the Module file was copied over from a much larger projects and there the all caps headers helps to navigate a larger file. However, the examples are rather trivial so there is not much value.

marvin-hansen commented on rust-tutorial/.bazelversion:
```diff
@@ -0,0 +1 @@
+USE_BAZEL_VERSION=7.2.0
```
Ok, Updated to Bazel 7.2.1.

marvin-hansen commented on rust-tutorial/.gitignore:
```diff
@@ -0,0 +1,3 @@
+*~
+/*/MODULE.bazel.lock
+/*/target/
```
Thanks for catching this one. 
This was an error. 
What should have been there is

/*/bazel-*

To exclude all generated Bazel ouput folders from all examples. That's fixed.

marvin-hansen commented on rust-tutorial/01-hello-clang/MODULE.bazel:
```diff
@@ -0,0 +1,55 @@
+    name = "llvm_toolchain",
+    llvm_versions = LLVM_VERSIONS,
+)
+use_repo(llvm, "llvm_toolchain", "llvm_toolchain_llvm")
```
Well, the purpose of the example is to show how to add a LLVM toolchain more or less as preview for the cross compile example. But your are right, adding some usage makes this more clear

marvin-hansen replied:
I thought about this and concluded that its best to remove the clang / llvm example
because it doesn't bring anything to the table that isn't covered in the other examples.

marvin-hansen commented on rust-tutorial/02-hello-cross/MODULE.bazel:
```diff
@@ -0,0 +1,94 @@
+)
+llvm.sysroot(
+    name = "llvm_toolchain_x86_with_sysroot",
+    label = "@@org_chromium_sysroot_linux_x64//:sysroot",
```
Excellent point. I've moved the sysroot into MODULE,bazel with the [use_repo_rule](https://bazel.build/rules/lib/globals/module#use_repo_rule)
Updated the cross & MUSL example.

marvin-hansen commented:
@Wyverald  Thank you for your valuable review. I've addressed the bulk of your suggestion. Still need to rework the llvm example, but please continue your review.

marvin-hansen commented:
Update:

1) Removed the LLVM example.
2) Added all Rust examples to CI
3) Updated Rust example Readme with a notice of supported platforms (Linux and Mac)

I've bumped into the max 80 tasks per config limit on the CI. Initially, I thought the the tutorial file was too big, so I've split that one into three (C++, Java, Rust) as suggested by the Ci error message, but as it turned out, it was the pre-submit file that imports everything and therefore exceeded the 80 task limit. In response, I've removed most Mac targets for the Rust examples except for cross compilation to stay within the limit and to give some more space to add more targets later. That's probably good since the mac targets tend to build slower.

Build is green. 

Please let me know if you want any further changes to any of the examples.

Wyverald commented on rust-tutorial/04-ffi/README.md:
```diff
@@ -0,0 +1,133 @@
+## Setup
+
+The setup is twofold, the Rust rules are declared in the MODULE file, 
+but the rules_cc are not yet available in the Bazelmod format and thus are declared in 
```
rules_cc is on BCR: https://registry.bazel.build/modules/rules_cc

It might be a bit out of date for your use case; if so, we can submit a newer version.

In general, though, rules_cc isn't really ready for public consumption. If you're just using C++ rules like `cc_import` and `cc_library`, I believe you can just use the natively bundled rules in Bazel (that is, don't `load` them).

Wyverald reviewed (approved):
I skimmed through the rest of the pages and saw no major concerns (other than the rules_cc thing). Thanks again for the very detailed examples!

marvin-hansen commented:
Thing is, while contributing to the rules_rust, I've discovered and reported a fairly hairy bug that causes the MUSL example to produce an [incorrect linked binary due to an incorrect toolchain resolution](https://github.com/bazelbuild/rules_rust/issues/2726). I discovered this bug after the examples were written during some routine structural integrity testing and that way I know that the MUSL example in this PR is also affected. The other examples were fine and especially the cross compilation examples were passing all tests. 

Point is, a fix is under way, but may take longer as it's not even merged so the bigger question is how to proceed?

I can think of two options:

1) Remove the MUSL example, review, merge, and wait for the fix to re-add the MUSL example later.
That's what we decided over at rules_rust as it's unclear when the fix lands.

2) Convert the PR into a draft, put it on hold until the MUSL fix arrives, apply the fix, and take it from there.

Thoughts?

alexeagle commented:
FYI @illicitonion

Wyverald commented:
> I can think of two options:

I'm okay with either option.

Also, please let me know when you've made all necessary changes to the PR and are ready to submit. I can override the CLA check from my end. (Alternatively, feel free to remove the "Co-authored-by" lines referring to me.)

marvin-hansen commented:
Update:

1)	Removed the WORKSPACE file with rules cc from example #4, FFI, and use the default rules. Also, updated the Readme.
2)	Updated all examples to the latest rules_rust (0.47.1)
3)	Renamed the rust-tutorial folder to rust-examples to better reflect the intention
4)	Removed the MUSL example for the time being.

I have to renovate the MUSL example anyways once a patch becomes available as the configuration changed, but more importantly I have to wait for the next rules_rust release that included the patch and I believe that’s going to take some time, therefore it’s sensible to remove the MUSL example for the time being and add it later once the rules_rust have been patched and the example has been updated. 

CI seems to fail only on Android and MacOS stuff, whereas the Rust examples all build.

Would be great if someone with more knowledge on Android / MacOS could take a look.

Please let me know if you want any further changes.

Otherwise, feel free to merge anytime when the CI issues are resolved.

Wyverald commented:
CI failures are unrelated (and already failing before this PR).

CLA check failure is a false positive -- somehow my personal email address is used there.

So I went ahead and merged this. Thank you again for your contribution!
